pkg/models/dashboard_queries.go

```diff
@@ -0,0 +1,23 @@
+package models
+
+import (
+	"github.com/grafana/grafana/pkg/components/simplejson"
+)
+
+func GetQueriesFromDashboard(dashboard *simplejson.Json) map[int64][]*simplejson.Json {
+	result := make(map[int64][]*simplejson.Json)
+
+	for _, panelObj := range dashboard.Get("panels").MustArray() {
+		panel := simplejson.NewFromAny(panelObj)
+
+		var panelQueries []*simplejson.Json
+
+		for _, queryObj := range panel.Get("targets").MustArray() {
+			panelQueries = append(panelQueries, simplejson.NewFromAny(queryObj))
+		}
+
+		result[panel.Get("id").MustInt64()] = panelQueries
+	}
+
+	return result
```

Comment:
this will not be enough, the data source is on the panel level (new dashboard schemas have data source on all queries, but not older ones)

So to know what data source the queries should target you need the data source on the panel level as well. 

Think it would be useful to return something like this

```go
struct DataQueryGroup {
  datasourceName string
  datasourceUid string 
  queries []DataQueryInfo
} 

struct DataQueryInfo {
  refId string
  datasourceName string
  datasourceUid string 
  model: *simplejson.Json
} 


```

Reply:
I like this structure as well. There may be some further optimizations or scheduling of queries based on datasources as well.

Reply:
As an intermediate step, could we pull in the datasource info from the panel and add it to the returned query? That is, make all queries look like new queries?

Reply:
Yes, if the query is missing a datasource you can take it from the panel level